src/app/domain/io.cozy.files/offlineFiles.ts

```diff
@@ -0,0 +1,115 @@
+import FileViewer from 'react-native-file-viewer'
+import RNFS from 'react-native-fs'
+
+import CozyClient from 'cozy-client'
+import type { FileDocument } from 'cozy-client/types/types'
+import Minilog from 'cozy-minilog'
+
+import {
+  addOfflineFileToConfiguration,
+  getOfflineFileFromConfiguration
+} from '/app/domain/io.cozy.files/offlineFilesConfiguration'
+import {
+  getDownloadUrlById,
+  getFileById
+} from '/app/domain/io.cozy.files/remoteFiles'
+import { getInstanceAndFqdnFromClient } from '/libs/client'
+import { normalizeFqdn } from '/libs/functions/stringHelpers'
+
+const log = Minilog('📁 Offline Files')
+
+export const getFilesFolder = (client: CozyClient): string => {
+  const { fqdn } = getInstanceAndFqdnFromClient(client)
+
+  const normalizedFqdn = normalizeFqdn(fqdn)
+
+  const destinationPath = `${RNFS.DocumentDirectoryPath}/${normalizedFqdn}/Files`
+
+  return destinationPath
+}
+
+const ensureFilesFolder = async (client: CozyClient): Promise<void> => {
+  await RNFS.mkdir(getFilesFolder(client))
+}
+
+export const downloadFile = async (
+  file: FileDocument,
+  client: CozyClient,
+  setDownloadProgress?: (progress: number) => void
+): Promise<string> => {
+  try {
+    log.debug('Download file', file._id)
+
+    await ensureFilesFolder(client)
+
+    const filesFolder = getFilesFolder(client)
+
+    const existingFile = await getOfflineFileFromConfiguration(file._id)
+
+    const remoteFile = await getFileById(client, file._id)
+
+    if (existingFile && existingFile.rev === remoteFile._rev) {
+      log.debug(`File ${file._id} already exist, returning existing one`)
+      return existingFile.path
+    }
+
+    const filename = remoteFile.name
+
+    const downloadURL = await getDownloadUrlById(client, file._id, filename)
+
+    const destinationPath = `${filesFolder}/${filename}`
+
+    const result = await RNFS.downloadFile({
+      fromUrl: downloadURL,
+      toFile: `${filesFolder}/${filename}`,
+      begin: () => undefined,
+      progress: res => {
+        const progressPercent = res.bytesWritten / res.contentLength
+        setDownloadProgress?.(progressPercent)
+      },
+      progressInterval: 100
+    }).promise
```

Comment:
Maybe extract the download logic like we did for the upload ? https://github.com/cozy/cozy-flagship-app/tree/master/src/app/domain/upload/services